### ISSUE TITLE
Unable to build with travis and hhvm after upgrading to phpunit 6.x

### ISSUE THREAD

**Issue by ottlinger** [sebastianbergmann/phpunit#2581]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 6.* / 6.0.10
| PHP version         | HipHop VM 3.6.6 (rel)
| Installation Method | Composer / PHAR

Unable to build with hhvm after upgrading to phpunit 6.x

Travis fails with
```
  Problem 1
    - sebastian/comparator 2.0.0 requires php ^7.0 -> your HHVM version does not satisfy that requirement.
    - sebastian/comparator 2.0.0 requires php ^7.0 -> your HHVM version does not satisfy that requirement.
    - Installation request for sebastian/comparator 2.0.0 -> satisfiable by sebastian/comparator[2.0.0].
```
after upgrading to latest 6.x phpunit and php7 as minimum:
https://travis-ci.org/ottlinger/hornherzogen/jobs/214279079

Is this a problem of composer with hhvm or phpunit with hhvm?

The build problems seem to be related to hhvm since builds gio through with other PHP versions.

**Comment by sebastianbergmann** [comment]
Sorry, but I cannot help you. PHPUnit is not supported on HHVM. Please open a ticket with the vendor of that alternative runtime for PHP.

**Comment by ottlinger** [comment]
@sebastianbergmann ups - ok. The older builds ran on HHVM .... interesting to know.

**Comment by k00ni** [comment]
The "problematic" part here is the following code:

```
if (version_compare('7.0.0', PHP_VERSION, '>')) {
```

which is located under [*phpunit/phpunit*](https://github.com/sebastianbergmann/phpunit/blob/master/phpunit).

I tried to "fool" PHPUnit by setting the environment variable **PHP_VERSION** to "7.0.1" in my .travis.yml:

```yaml
matrix:
    fast_finish: true
    include:
        - php: hhvm
          env: PHP_VERSION=7.0.1

script: PHP_VERSION=7.0.1 vendor/bin/phpunit
```
But unfortunately it doesn't work. Why forcing a certain PHP version in the script if you already require it in composer.json?

**Comment by sebastianbergmann** [comment]
Because Composer is not the only supported method of installing PHPUnit and because it is common to have multiple versions of PHP installed.

**Comment by k00ni** [comment]
Ok, i understand.

But are there any known bugs, problems or incompatibilities with HHVM, which prevent supporting it?

**Comment by ottlinger** [comment]
:+1: the interesting thing is that builds seemed to work fine with older PHPUnit versions .... maybe there is a more generic solution that integrates smoothly with Travis/Github.

**Comment by fredemmott** [comment]
You want `hhvm.php7.all=1` in your php.ini, for both composer and PHPUnit. I've not actually tested running the versions of phpunit that require PHP7 though, but that's how you make it pass those tests (and switch the runtime behavior).

This wont' work on HHVM 3.6.6, which has been entirely unsupported for over a year, and has known security vulnerabilities; you need to upgrade.

**Comment by fredemmott** [comment]
If you can't figure out where the php.ini is that travis cares about, `hhvm -d hhvm.php7.all=1`will also work

**Comment by fredemmott** [comment]
So, that makes composer happy with the version check, but composer doesn't actually work with it. That's an HHVM bug, I'll poke the team...

**Comment by k00ni** [comment]
@fredemmott: thanks for the tip, but i am struggling. 

My .travis.yml has in the `before_install` section the following to new lines:

```yaml
before_install:
   # [...]
    - echo "hhvm.php7.all=1" >> /tmp/hhvm.php.ini
    - phpenv config-add /tmp/hhvm.php.ini
```

but the [build](https://travis-ci.org/k00ni/knorke/jobs/229695548) is still failing, because PHPUnit quits with:

```
This version of PHPUnit is supported on PHP 7.0 and PHP 7.1.
You are using PHP 5.6.99-hhvm (/usr/bin/hhvm).
```
Any idea?

**Comment by TysonAndre** [comment]
Running phpunit 6 on hhvm won't work until https://github.com/facebook/hhvm/issues/6037 is fixed.

```diff
--- a/phpunit
+++ b/phpunit
@@ -8,8 +8,10 @@
  * For the full copyright and license information, please view the LICENSE
  * file that was distributed with this source code.
  */
-
-if (version_compare('7.0.0', PHP_VERSION, '>')) {
+// Optimistically hope that new hhvm releases are compatible, but 3.11.1 didn't work
+// https://github.com/facebook/hhvm/issues/6037 still doesn't have a solution
+if (version_compare('7.0.0', PHP_VERSION, '>') &&
+        !(defined('HHVM_VERSION') && VERSION_COMPARE(HHVM_VERSION, '3.11.0', '>=') && ini_get('hhvm.php7.all'))) {
```


Error that was seen:

```
Runtime:       HHVM 3.11.1
Configuration: /phpunit-src/phpunit.xml

.
Fatal error: Uncaught TypeError: Argument 1 passed to PHPUnit\Framework\TestCase::onNotSuccessfulTest() must be an instance of Throwable, PHPUnit\Framework\ExpectationFailedException given in /phpunit-src/src/Framework/TestCase.php:2133
Stack trace:
#0 /phpunit-src/src/Framework/TestCase.php(1037): PHPUnit\Framework\TestCase->onNotSuccessfulTest()
#1 /phpunit-src/src/Framework/TestResult.php(696): PHPUnit\Framework\TestCase->runBare()
#2 /phpunit-src/src/Framework/TestCase.php(883): PHPUnit\Framework\TestResult->run()
#3 /phpunit-src/src/Framework/TestSuite.php(746): PHPUnit\Framework\TestCase->run()
#4 /phpunit-src/src/Framework/TestSuite.php(746): PHPUnit\Framework\TestSuite->run()
#5 /phpunit-src/src/Framework/TestSuite.php(746): PHPUnit\Framework\TestSuite->run()
#6 /phpunit-src/src/TextUI/TestRunner.php(537): PHPUnit\Framework\TestSuite->run()
#7 /phpunit-src/src/TextUI/Command.php(210): PHPUnit\TextUI\TestRunner->doRun()
#8 /phpunit-src/src/TextUI/Command.php(141): PHPUnit\TextUI\Command->run()
#9 /phpunit-src/phpunit(55): PHPUnit\TextUI\Command::main()
```

**Comment by fredemmott** [comment]
@TysonAndre : that issue should have been closed; I've not checked phpunit yet (I'm looking at composer first), but if you find any problems with Throwable on 3.20+ that aren't https://github.com/facebook/hhvm/issues/7837 , please file a new issue against HHVM. Also, 3.11 is *very* old and unsupported.

**Comment by fredemmott** [comment]
@k00ni:

> Running hhvm -d hhvm.php7.all=1 in the PHP environment leads to an error

Please be more specific about 'an error', and this should really be reported against HHVM, not PHPUnit.

Also, sadly, HHVM + PHP7 isn't really in a workable state right now - the biggest known problems were fixed recently, but that just uncovered new ones. I'm back working on HHVM for a month, trying to get both composer and PHPUnit working again... https://github.com/facebook/hhvm/issues/7869 is the high-level issue for me to fix composer. Until a release with that fixed, things basically aren't going to work.

**Comment by k00ni** [comment]
@fredemmott:

Can you please paste a one-liner which allows me to run `hhvm -d hhvm.php7.all=1` inside Travis when using HHVM, but not when in PHP mode? I don't want to crash my existing PHP  pipelines by running a HHVM command. Thanks in advance.

**Comment by TysonAndre** [comment]
You're correct, I was running an outdated version of hhvm and didn't notice. I was running 3.11. 3.19 is out, and has better support for Throwable.

I was trying to get it to work just now, but there are still minor issues.

That patch I mentioned is unnecessary now, `PHP_VERSION` becomes 7.1.0-...

**Comment by fredemmott** [comment]
@k00ni: doesn't PHP just ignore unrecognized INI settings? Why not always set it? Also, none of the versions of HHVM available on Travis are going to work well in PHP7 mode, and this is unlikely to change until Travis support sudo mode on Ubuntu 16.04 as recent versions of HHVM won't run on 14.04.

**Comment by TysonAndre** [comment]
https://docs.travis-ci.com/user/docker/ may be a workaround (Use ubuntu 16.04 or hhvm image derived from that in a Dockerfile), may slow down builds.

**Comment by fredemmott** [comment]
Here's an HHVM-only example: https://github.com/facebook/xhp-lib/blob/master/.travis.yml
 + https://github.com/facebook/xhp-lib/blob/master/.travis.sh

Sorry, I don't have a PHP7 + HHVM example, and I'm not planning on looking into that until HHVM can run composer and phpunit properly in php7 mode.

**Comment by k00ni** [comment]
> [...] until HHVM can run composer and phpunit properly in php7 mode.

:+1: Looking forward to it

**Comment by fredemmott** [comment]
- composer appears fixed in HHVM master (expecting to backport fixes to 3.20.3): https://github.com/facebook/hhvm/issues/7869
 - given composer works, I've started looking at PHPUnit. https://github.com/facebook/hhvm/issues/7876 - I'm not expecting to post any more replies to this issue.
